tests/cpp/test_multidevice_pipeline.cpp

```diff
@@ -157,24 +188,22 @@ TEST_P(PipelineTestTwoStages, Communication) {
     mesh1 = mesh0;
   }
 
-  int first_axis_extent = 3;
+  std::vector<int64_t> unsharded_input_sizes = {3, 2, 3, 5};
   if (is_stage0_sharded) {
-    first_axis_extent = mesh0.vector().size();
-  } else if (is_stage1_sharded) {
-    first_axis_extent = mesh1.vector().size();
+    unsharded_input_sizes[sharded_dim] = mesh0.vector().size();
```

Comment:
I don't understand this sentence. We are replacing the extent of an axis with a mesh rank? They don't seem to be logically connected. 

(This is for my education and not a problem with a PR. The old code is as mysterious to me :)

Reply:
It's replacing the extent of the exist with the extent of the mesh.  We have a constraint that the extent of a device parallel axis == number of devices.
When we support split it will be less clunky.

Reply:
> I don't understand this sentence. We are replacing the extent of an axis with a mesh rank? They don't seem to be logically connected.

The axis and device mesh extent are logically eauqlly. It is exactly the same as binding some axis extent to gridDim.x or blockDim.x. But here block/grid is replaced by the device mesh. Does it make sense ?